kasa/discover.py

```diff
@@ -224,11 +228,25 @@ async def do_discover(self) -> None:
         _LOGGER.debug("[DISCOVERY] %s >> %s", self.target, Discover.DISCOVERY_QUERY)
         encrypted_req = XorEncryption.encrypt(req)
         sleep_between_packets = self.discovery_timeout / self.discovery_packets
+
+        key_payload = {"params": {"rsa_key": self.keypair.get_public_pem().decode()}}
+
+        encoded_payload = json_dumps(key_payload).encode()
+        version = 2
+        nonce = 1337
+        initial_crc = 0x5A6B7C8D  # https://labs.withsecure.com/advisories/tp-link-ac1750-pwn2own-2019
+        disco_header = struct.pack(
+            ">BBHHBBII", version, 0, 1, len(encoded_payload), 17, 0, nonce, initial_crc
+        )
+
+        query_2 = bytearray(disco_header + encoded_payload)
+        query_2[12:16] = binascii.crc32(query_2).to_bytes(length=4, byteorder="big")
         for _ in range(self.discovery_packets):
             if self.target in self.seen_hosts:  # Stop sending for discover_single
                 break
             self.transport.sendto(encrypted_req[4:], self.target_1)  # type: ignore
             self.transport.sendto(Discover.DISCOVERY_QUERY_2, self.target_2)  # type: ignore
+            self.transport.sendto(query_2, self.target_2)  # type: ignore
```

Comment:
IIRC, the devices that do not require the key as part of the payload will still answer to queries when the key is there, meaning that we could probably do with a single request here? Worth checking out at least.

Reply:
Fine by me :+1: